internal/exec/runner.go

```diff
@@ -0,0 +1,922 @@
+// as an error with a non-zero exit code, seems inconsistent, this needs refactoring
+
+// filename is optional
+// if set, it will update the Failures to have this filename
+// will be sorted
+func (r *runner) printFailures(filename string, meta *meta, failures ...*text.Failure) error {
+	for _, failure := range failures {
+		if filename != "" {
+			failure.Filename = filename
+		}
+	}
+	failureFields, err := text.ParseColonSeparatedFailureFields(r.printFields)
+	if err != nil {
+		return err
+	}
+	text.SortFailures(failures)
+	bufWriter := bufio.NewWriter(r.output)
+	for _, failure := range failures {
+		shouldPrint := false
+		if meta.InDirModeSingleFilename == "" || meta.InDirModeSingleFilename == failure.Filename {
+			shouldPrint = true
+		} else if meta.InDirModeSingleFilename != "" {
+			// TODO: the compiler may not return the rel path due to logic in bestFilePath
+			absSingleFilename, err := absClean(meta.InDirModeSingleFilename)
+			if err != nil {
+				return err
+			}
+			absFailureFilename, err := absClean(failure.Filename)
+			if err != nil {
+				return err
+			}
+			if absSingleFilename == absFailureFilename {
+				shouldPrint = true
+			}
+		}
+		if shouldPrint {
+			if r.harbormaster {
+				harbormasterLintResult, err := phab.TextFailureToHarbormasterLintResult(failure)
+				if err != nil {
+					return err
+				}
+				data, err := json.Marshal(harbormasterLintResult)
+				if err != nil {
+					return err
+				}
+				if _, err := fmt.Fprintln(bufWriter, string(data)); err != nil {
+					return err
+				}
+			} else if err := failure.Fprintln(bufWriter, failureFields...); err != nil {
+				return err
+			}
+		}
+	}
+	return bufWriter.Flush()
+}
+
+func (r *runner) printLinters(linters []lint.Linter) error {
+	sort.Slice(linters, func(i int, j int) bool { return linters[i].ID() < linters[j].ID() })
+	tabWriter := newTabWriter(r.output)
+	for _, linter := range linters {
+		if _, err := fmt.Fprintf(tabWriter, "%s\t%s\n", linter.ID(), linter.Purpose()); err != nil {
+			return err
+		}
+	}
+	return tabWriter.Flush()
+}
+
+func (r *runner) printAffectedFiles(meta *meta) {
+	for _, protoSet := range meta.ProtoSets {
+		for _, files := range protoSet.DirPathToFiles {
+			for _, file := range files {
+				r.logger.Debug("using file", zap.String("file", file.DisplayPath))
+			}
+		}
+	}
+}
+
+func (r *runner) println(s string) error {
+	if s == "" {
+		return nil
+	}
+	_, err := fmt.Fprintln(r.output, s)
+	return err
+}
+
+func (r *runner) getInputData(arg string) ([]byte, error) {
+	if arg == "-" {
+		return ioutil.ReadAll(r.input)
+	}
+	return []byte(arg), nil
+}
+
+func (r *runner) getInputReader(arg string) io.Reader {
+	if arg == "-" {
+		return r.input
+	}
+	return bytes.NewReader([]byte(arg))
+}
+
+func newExitErrorf(code int, format string, args ...interface{}) *ExitError {
+	return &ExitError{
+		Code:    code,
+		Message: fmt.Sprintf(format, args...),
+	}
+}
+
+// TODO: this is copied in three places
```

Comment:
We could put this in the `strs` directory, not a perfect spot, but better than repeating

Reply:
`strs` no, you wouldn't put this in `https://golang.org/pkg/strings`, which is what `strs` corresponds to. We can make a new package, filed https://github.com/uber/prototool/issues/133.